Add guest_email to checkout view

crgwbr opened the pull request:
Add guest_email field to checkout view. This is needed for anonymous checkouts.

specialunderwear commented:
The tests are failing, please fix the unit tests that don't count on this new field.

crgwbr commented:
OK, back to you @specialunderwear. I added some additional tests as for anonymous checkouts.

specialunderwear commented on oscarapi/tests/testcheckout.py:
```diff
@@ -176,16 +179,56 @@ def test_checkout_header(self):
+            }
+        }
+
+        with mock.patch('django.conf.settings.OSCAR_ALLOW_ANON_CHECKOUT', True, create=True):
```
Why not:

with self.settings(OSCAR_ALLOW_ANON_CHECKOUT=True):

maerteijn replied:
+1 on that. Depending on `mock>=1.3.0` just for this already build-in functionality is not really needed.

crgwbr replied:
Huh, I've never seen that before. Will change now

specialunderwear commented on setup.py:
```diff
@@ -36,7 +36,8 @@
         'django-oscar>=1.1',
-        'djangorestframework>=3.1.0'
+        'djangorestframework>=3.1.0',
+        'mock>=1.3.0',
```
Not needed? Or should be in test extras

maerteijn commented on oscarapi/serializers/checkout.py:
```diff
@@ -182,6 +182,7 @@ class Meta:
 class CheckoutSerializer(serializers.Serializer, OrderPlacementMixin):
     basket = serializers.HyperlinkedRelatedField(
         view_name='basket-detail', queryset=Basket.objects)
+    guest_email = serializers.EmailField(allow_blank=True, required=False)
```
Why would you add `allow_blank=True`? Now it's possible to do an anonymous checkout without a valid email address

maerteijn commented on oscarapi/serializers/checkout.py:
```diff
@@ -195,11 +196,21 @@ def get_initial_order_status(self, basket):
+            if 'guest_email' in attrs:
+                # Don't store guest_email field if the user is authenticated
+                attrs['guest_email'] = None
+
```
I would say to write it like this:
```python
        if request.user.is_anonymous():
            if not settings.OSCAR_ALLOW_ANON_CHECKOUT:
                message = _('Anonymous checkout forbidden')
                raise serializers.ValidationError(message)

            if 'guest_email' not in attrs:
                # always require the guest email field if the user is anonymous
                message = _('Guest email is required for anonymous checkouts')
                raise serializers.ValidationError(message)
        else:
            # delete guest_email when it's there because we are authenticated
            if 'guest_email' in attrs:
                del attrs['guest_email']
```

crgwbr commented:
@maerteijn ok I addressed your concerns. I've added the allow blank param because some JS frameworks will send the param as an empty string and the only time that's actually an issue is when the user is anonymous. The validation logic now reflects this.

maerteijn commented:
Looks a lot cleaner and clearer now. The only think that's missing in`def test_anonymous_checkout` is to test for an empty (or not specified) `guest_email` address. So I would say that you remove the `guest_email` from the request dictionary and write something like this:

``` python
with self.settings(OSCAR_ALLOW_ANON_CHECKOUT=True):
    response = self.post('api-checkout', **request)
    self.assertEqual(response.status_code, 406)
    response = self.post('api-basket-add-product', url="http://testserver/api/products/1/", quantity=5)
    self.assertEqual(response.status_code, 200)

    # no guest email specified should say 406
    response = self.post('api-checkout', **request)
    self.assertEqual(response.status_code, 406)

    # an empty email address should say this as well
    request['guest_email'] = ''
    response = self.post('api-checkout', **request)
    self.assertEqual(response.status_code, 406)

    # a valid email address should say 200
    request['guest_email'] = 'henk@example.com'
    response = self.post('api-checkout', **request)
    self.assertEqual(response.status_code, 200)
    self.assertEqual(response.data['guest_email'], 'henk@example.com')
    self.assertEqual(Basket.objects.get(pk=basket_id).status, 'Frozen', 'Basket should be frozen after placing order and before payment')

bufke commented:
@maerteijn - @crgwbr asked me to finish this up for him. I made the changes pretty as you suggested. Anything else needed?

specialunderwear commented:
Is ok with me

maerteijn commented:
Fine by me too.
